### PR TITLE
Websockets

### PR DESCRIPTION
**Pull request by daurnimator**
- [x] framing functions
  - [x] client side
  - [x] server side
  - [x] documentation
  - [x] example
  - [x] tests

### CODE REVIEW

File: http/websocket.lua [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+		length_extra = ""
+	elseif b2 <= 0xFFFF then -- 2-byte length
+		b2 = 126
+		length_extra = spack(">I2", #data)
+	else -- 8-byte length
+		b2 = 127
+		length_extra = spack(">I8", #data)
+	end
+
+	local key = ""
```
Comment (daurnimator) [inline http/websocket.lua]: Remove

**Comment by ChickenNuggers** [comment]
```
22:54:30 <~daurnimator> the server should be the same :send, :read, :close [as the client]
```

**Comment by daurnimator** [comment]
I just renamed :read to :receive. See the commit message for more rationale.

**Comment by daurnimator** [comment]
API question: How should error codes be returned to the caller.
e.g. you call `:receive`; the peer causes a in-frame protocol error: you have multiple error codes to pick between:
 1. the code you send *to* the peer informing them of their error
 2. the code you *receive* in their acknowledgement of the error
 3. a generated `errno`

`2.` is currently stored in `ws.got_close_code`, so I've tentatively picked option 1.

**Pull request merged**